pullrequests/reverse_linked_list/step3.go

```diff
@@ -0,0 +1,47 @@
+//lint:file-ignore U1000 Ignore all unused code
+package reverselinkedlist
+
+/*
+	時間: 2分、2分
+	イテレーティブな方法と再帰を使った方法の２通りで解きました。
+
+	質問：
+		- 20行目について、１行にまとめているのですが、読みやすさなどの点ではどう思いますか？
+
+		- 与えられたhead自体を操作するのは避けた方が良いか？（例えばcurrなどの別の変数に代入して
+			から操作した方が良いか）
+
+		- 変数名としてreversedHeadか、reversedListHeadのどちらの方が良いか？
+			- reversedListHeadだとこれぐらいのコードの短さで使うにしては少し長すぎるかなと思った
+				のでここではreversedHeadでも十分通じると思い、これにしました
+
+	他の人のPRを見て思ったこと
+		- スタックを使って解く方法も確かにあるなと思いました。パフォーマンス的には再帰を使って解く
+			方法とほぼ同じ？（再帰もスタックを使っていると言えば使っている）
```

Comment:
この点、ちゃんと言語化出来るようになってると良いと思います。個人的にですが、エンジニアとして一緒に働くなら、時間計算量、空間計算量について考えられるのは最低限出来ていて欲しいかなと思うのが理由になります。
実際面接でもこのあたりの説明が求められることが多いように感じます。

Reply:
オーダー記法的にはパフォーマンスは変わらないですが、その他の点でパフォーマンスに違いが出たりしますか？というニュアンスの質問でした、言葉足らずですいません、、

Reply:
あ、なるほどです。こちらこそ意図が汲み取れて無くて失礼しました🙇‍♂️
その点についてコメントすると、ぱっと思いつく限り、スタックを使う場合はそのインスタンスがヒープ領域に置かれ、再帰を使う場合はスタック領域に計算に必要な値が積まれるという違いがあるなと思いました。一般的に後者の領域のほうが早く効率的にアクセス出来るので、再帰のほうが処理が早くなりそうです (LeetCode のテストケース程度の大きさのデータを扱う上ではあまり差は出ないかもしれませんが...)。

Reply:
その視点は確かにと思ったのですが、Goのドキュメントを見てみると、局所的に使われていて、サイズも大きくない場合は大抵の場合はスタックに確保されてそうです。Goの動的配列（スライス）は内部的には配列が使われていてそれに対するポインタ情報を持たせて機能的に拡張しているだけなので、同様に大抵の場合はスタックに領域が確保されるのかなと思います。

ただサイズが大きくなってくるとヒープに確保されるようになるのかなと思うので、スタックオーバーフローの危険性がない分、スライスを使ったスタックを使う方が良いのではと思ったりしました。

また、再帰呼び出しを使うと呼び出しごとに関数の戻りアドレスやローカル変数などのデータを持ったスタックフレームを作って積んでいく必要があるので、単にスライスを使って要素を追加・削除していく形で実装した方がオーバーヘッドが小さいのではと今思ったりしました。

> From a correctness standpoint, you don’t need to know. Each variable in Go exists as long as there are references to it. The storage location chosen by the implementation is irrelevant to the semantics of the language.
> 
> The storage location does have an effect on writing efficient programs. When possible, the Go compilers will allocate variables that are local to a function in that function’s stack frame. However, if the compiler cannot prove that the variable is not referenced after the function returns, then the compiler must allocate the variable on the garbage-collected heap to avoid dangling pointer errors. Also, if a local variable is very large, it might make more sense to store it on the heap rather than the stack.
> 
> In the current compilers, if a variable has its address taken, that variable is a candidate for allocation on the heap. However, a basic escape analysis recognizes some cases when such variables will not live past the return from the function and can reside on the stack.

https://go.dev/doc/faq#stack_or_heap

Reply:
これは要調査ですね。

> また、再帰呼び出しを使うと呼び出しごとに関数の戻りアドレスやローカル変数などのデータを持ったスタックフレームを作って積んでいく必要があるので、単にスライスを使って要素を追加・削除していく形で実装した方がオーバーヘッドが小さいのではと今思ったりしました。